### ISSUE TITLE
Thoughts on parallel computation

### ISSUE THREAD

**Issue by plajjan** [actonlang/acton#394]
I think we have the opportunity for supporting two types of parallel computation; implicit and explicit. The explicit kind is quite straight forward; start multiple actors and distribute the work across them. There might be room for certain code patterns and maybe even built-in support for doing load balancing of work.

Implicit parallelism is much more exciting. It is when we can determine that a certain operation, like a list comprehension which is not explicitly coded for parallelism, can be executed in parallel.

How should we run those in parallel?

The current RTS design is to use one worker thread per CPU core/thread. Each thread picks up a continuation and runs its. A parallelizable computation would be a section of a continuation and since a continuation runs on a single worker thread, we do not naturally have a context in which we can execute something in parallel. What to do? Should we turn the continuation into some form of continuation that is scheduled across multiple worker threads? We don't want that allocation to be static, like we could perhaps determine already at compile time that a certain operation can be run in parallel, but we cannot decide for N parallelism, since we do not know how many worker threads are available to us at run time. It must be dynamic! Also, maybe the other worker threads are busy, in which case it is probably better to execute on one core than to wait for others to become available.

We could just work with the current continuation approach and so whenever when we encounter a parallel computation, we split it up as smaller continuations blocks and place on the readyQ for other threads (and maybe our own thread too) to pick up. We probably shouldn't add those jobs at the end of the queue though? We'd have to add at the beginning or so? Also tricky to determine size of a job, like we don't want to add a gazillion very small jobs to the queue, then the overhead of adding and picking them up from the queue far outweighs any gain from running the actual work in parallel. Figuring out the size / cost of a computation is a classic problem. Maybe we shouldn't mix this stuff up with the readyQ but rather have another Q? Is that better? Though then what do we then wait on? Hmm, or do we just do parallel computation when the readyQ is empty, indicating that we have available capacity?

Another approach is to not pin worker threads to CPU threads, so we could let a worker thread maybe start other computation threads for dealing with a parallel operation. This would be a rather big step away from what we have thus seen as pretty much an axiom and a good design for a performant system. Doesn't feel good.

This is by no means urgent, but it might perhaps be good to keep it in mind if we do any other modifications to the RTS.
